macros/src/main/scala/MacroCompiler.scala

```diff
@@ -0,0 +1,635 @@
+                        libs: Option[Seq[Macro]],
+                        costMetric: CostMetric = CostMetric.default,
+                        mode: MacroCompilerAnnotation.CompilerMode = MacroCompilerAnnotation.Default) extends firrtl.passes.Pass {
+  def compile(mem: Macro, lib: Macro): Option[(Module, ExtModule)] = {
+    val pairedPorts = mem.sortedPorts zip lib.sortedPorts
+
+    // Parallel mapping
+
+    /**
+     * This is a list of submemories by width.
+     * The tuples are (lsb, msb) inclusive.
+     * e.g. (0, 7) and (8, 15) might be a split for a width=16 memory into two
+     * width=8 memories.
+     */
+    val bitPairs = ArrayBuffer[(BigInt, BigInt)]()
+    var currentLSB = 0
+
+    // Process every bit in the mem width.
+    for (memBit <- 0 until mem.src.width) {
+      val bitsInCurrentMem = memBit - currentLSB
+
+      /**
+       * Helper function to check if it's time to split memories.
+       * @param effectiveLibWidth Split memory when we have this many bits.
+       */
+      def splitMemory(effectiveLibWidth: Int): Unit = {
+        if (bitsInCurrentMem == effectiveLibWidth) {
+          bitPairs += ((currentLSB, memBit - 1))
+          currentLSB = memBit
+        }
+      }
+
+      for ((memPort, libPort) <- pairedPorts) {
+
+        // Make sure we don't have a maskGran larger than the width of the memory.
+        assert (memPort.src.effectiveMaskGran <= memPort.src.width)
+        assert (libPort.src.effectiveMaskGran <= libPort.src.width)
+
+        val libWidth = libPort.src.width
+
+        // Don't consider cases of maskGran == width as "masked" since those masks
+        // effectively function as write-enable bits.
+        val memMask = if (memPort.src.effectiveMaskGran == memPort.src.width) None else memPort.src.maskGran
+        val libMask = if (libPort.src.effectiveMaskGran == libPort.src.width) None else libPort.src.maskGran
+
+        (memMask, libMask) match {
+          // Neither lib nor mem is masked.
+          // No problems here.
+          case (None, None) => splitMemory(libWidth)
+
+          // Only the lib is masked.
+          // Not an issue; we can just make all the bits in the lib mask enabled.
+          case (None, Some(p)) => splitMemory(libWidth)
+
+          // Only the mem is masked.
+          case (Some(p), None) => {
+            if (p % libPort.src.width == 0) {
+              // If the mem mask is a multiple of the lib width, then we're good.
+              // Just roll over every lib width as usual.
+              // e.g. lib width=4, mem maskGran={4, 8, 12, 16, ...}
+              splitMemory(libWidth)
+            } else if (libPort.src.width % p == 0) {
+              // Lib width is a multiple of the mem mask.
+              // Consider the case where mem mask = 4 but lib width = 8, unmasked.
+              // We can still compile, but will need to waste the extra bits.
+              splitMemory(memMask.get)
+            } else {
```

Comment:
I'm a bit hesitant to merge them directly since I'd like to guard against the `!(p < libPort.src.width)` case and I would prefer to keep them separate (at least I find it easier to read) as opposed to combining them into a else if (... & ...) clause above, though I'm happy to hear other thoughts on this.